azurelinuxagent/ga/exthandlers.py

```diff
@@ -708,6 +718,70 @@ def report_event(self, message="", is_success=True, duration=0):
         add_event(name=self.ext_handler.name, version=ext_handler_version, message=message,
                   op=self.operation, is_success=is_success, duration=duration)
 
+    def try_get_pkg_file(self):
+        """
+        Attempt to get the full name of an extension ZIP package that was downloaded
+        based on the name and the version of the extension handler.
+
+        This is not a simple process. The extension name in goal state combines
+        namespace and type, but the version is separate. When the agent downloads
+        the extension from storage, the name separates namespace, type, and version.
+
+        This means the code has to figure out how to divide a combined namespace
+        and type. This is not hard, but the inconsistency between systems is
+        annoying.
+
+        Here are the naming conventions used in different places.
+
+        Goal State:
+          name="<Namespace>.<Type>" version="1.0.0"
+
+        URL:
+          "<Namespace>__<Type>__<Version>"
+
+        The code has to determine how to break namespace and type to properly
+        compare the Goal State value with the URL value. The approach taken by
+        the code below is to normalize the value to Goal State's view of the
+        world.
+
+        Given a namespace of Microsoft.OSTCExtensions, a typeof CustomScriptForLinux,
+          and a Version of 1.5.2.2
+        Then the method will determine that
+            name="Microsoft.OSTCExtensions.CustomScriptForLinux" version="1.5.2.2"
+          is equivalent to
+            "Microsoft.OSTCExtensions__CustomScriptForLinux__1.5.2.2
+          by normalizing both values to
+            "Microsoft.OSTCExtensions.CustomScriptForLinux.1.5.2.2
+        And that the full path to the extension's archive is
+            "/var/lib/waagent/Microsoft.OSTCExtensions.CustomScriptForLinux.1.5.2.2.zip"
+
+        :return: full path to package's .zip or None
+        :rtype: str
+        """
+        if self.pkg_file is not None:
+            return self.pkg_file
+
+        def normalize_ext_name(f):
+            parts = f.split('__')
+            if len(parts) != 3:
+                return None
+            return "{0}.{1}.{2}".format(parts[0], parts[1], parts[2])
```

Comment:
Consider `return ".".join(parts)` as a clearer expression of what you're doing (dot-connect these three elements). Both work, and I can live with what you wrote.

To be honest, the right answer is to refactor the implementation into a static method or simple function that takes an extension publisher, an extension type name, and a version, and returns the canonicalized string.

Reply:
I was trying to find a better way to express line 768, and `return ".".join(parts)` is what I want.

I would take it one step further in a future commit.  It deserves a proper type to encase these behaviors. The fact that PIR and CRP cannot agree on a consistent naming conventions stinks, but that's life.  This is not a string this is a type called ExtensionName (or some such).